Shell/AST.cpp

```diff
@@ -190,12 +190,17 @@ Node::~Node()
 void And::dump(int level) const
 {
     Node::dump(level);
-    m_left->dump(level + 1);
-    m_right->dump(level + 1);
+    if (m_left)
+        m_left->dump(level + 1);
+    if (m_right)
+        m_right->dump(level + 1);
 }
 
 RefPtr<Value> And::run(RefPtr<Shell> shell)
 {
+    ASSERT(m_left);
+    ASSERT(m_right);
```

Comment:
Instead of this, `m_left` and `m_right` should be `NonnullRefPtr`s :)

Reply:
Previously, this made sense, as the parser wouldn't construct half-correct nodes
but now that the actual structure is needed (and examined) via the formatter, these may well be null.

I don't think `NonnullRefPtr` would work in general.

Though in the case of `And`, I believe the parser does substitute a `SyntaxError` node if the rhs is null, so it applies in this specific case :thinking:

Reply:
I'll try making them all `NonnullRefPtr`'s and putting `SyntaxError` nodes inside them, let's be a bit more hygienic~

Reply:
Grand!

Reply:
That was a workout and a half :P